app/controllers/application_controller.rb

```diff
@@ -1,10 +1,151 @@
+
+  get '/' do
+    erb :'/index'
+  end
+
+  get '/signup' do
+    if !session[:user_id]
+      erb :'users/create_user', locals: {message: "Please sign up before you sign in"}
+    else
+      redirect to '/tweets'
+    end
+  end
+
+  post '/signup' do
+    if params.any? { |k, v| v.length <=0 }
+      redirect to '/signup'
+    else
+      @user = User.new(:username => params[:username], :email => params[:email], :password => params[:password])
+      @user.save
+      session[:user_id] = @user.id
+      redirect to '/tweets'
+    end
+  end
+
+  get '/login' do
+    if !session[:user_id]
+      erb :'users/login'
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(:username => params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect to '/signup'
+    end
+  end
+
+  get '/logout' do
+    if !!session[:user_id]
+      session.destroy
+      redirect to '/login'
+    else
+      redirect to '/'
+    end
   end
 
-end
\ No newline at end of file
+  get '/tweets' do
+      if session[:user_id]
+        @tweets = Tweet.all
+        erb :'tweets/tweets'
+      else
+        redirect to '/login'
+      end
+    end
+
+    get '/tweets/new' do
+      if session[:user_id]
+        erb :'tweets/create_tweet'
+      else
+        redirect to '/login'
+      end
+    end
+
+    post '/tweets' do
+      if params[:content] == ""
+        redirect to "/tweets/new"
+      else
+        @tweet = Tweet.create(:content => params[:content], :user_id => current_user.id)
+        redirect to "/tweets/#{@tweet.id}"
+      end
+    end
+
+    get '/tweets/:id' do
+      if session[:user_id]
+        @tweet = Tweet.find(params[:id])
+      #  binding.pry
+        erb :'tweets/show_tweet'
+      else
+        redirect to '/login'
+      end
+    end
+
+    get '/tweets/:id/edit' do
+      if session[:user_id]
+        @tweet = Tweet.find(params[:id])
+        if @tweet.user_id == session[:user_id]
+          erb :'tweets/edit_tweet'
+        else
+          redirect to '/tweets'
+        end
+      else
+        redirect to '/login'
+      end
+    end
+
+    post '/tweets/:id' do
+      @tweet = Tweet.find(params[:id])
+      if params[:content] == ""
+        redirect to "/tweets/#{@tweet.id}/edit"
+      else
+        @tweet.content = params[:content]
+        @tweet.save
+        redirect to "/tweets/#{@tweet.id}"
+      end
+    end
+
+    get '/users/:slug' do
+      @user = User.find_by_slug(params[:slug])
+      erb :'users/show'
+    end
+
+    delete '/tweets/:id/delete' do
+      if session[:user_id]
+        @tweet = Tweet.find(params[:id])
+        if @tweet.user_id == session[:user_id]
+          @tweet.delete
+          redirect to '/tweets'
+        else
+          redirect to '/tweets'
+        end
+      else
+        redirect to '/login'
+      end
+    end
+
+
+    helpers do
+       def logged_in?
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```